### ISSUE TITLE
Throttling causing final event to never be fired

### ISSUE THREAD

**Issue by Jimbly** [paulmillr/chokidar#1084]
**Describe the bug**

If a file causes multiple low-level change events, a final event is never fired after the final change event, causing an application using chokidar to never see the final, updated file.

**Versions (please complete the following information):**
 - Chokidar version 3.5.1
 - Versions tested: Node v14.16.0 and v12.18.1 and Ubuntu 16.04
 - Versions tested: Node v12.18.1 on Windows 10

**To Reproduce:**

Write a file in two chunks using `fs.createWriteStream().write().write().end()`, with a slight delay between the writes.  A watching program that listens for change events and reads the file will never see the final file.

```js
const assert = require('assert');
const chokidar = require('chokidar');
const fs = require('fs');
let watcher = chokidar.watch('.', {
  cwd: __dirname,
  ignoreInitial: true,
});
const filename = `${__dirname}/test.txt`;
let data;
let start = Date.now();
function log(...args) {
  console.log(Date.now() - start, ...args);
}
function onChange(file) {
  log(`Event on ${file}`);
  data = fs.readFileSync(filename, 'utf8');
  log(`Read ${data}`);
}
watcher.on('add', onChange);
watcher.on('change', onChange);
watcher.on('raw', (event, path, details) => {
  log('Raw event info:', event, path, details);
});
watcher.on('ready', () => {
  log('Ready');
  let out = fs.createWriteStream(filename);
  out.write('first');
  setTimeout(() => { // On Windows, reproduces even without the setTimeout
    out.write('second');
    out.end();
    setTimeout(() => {
      if (data !== 'firstsecond') {
        log('Never saw final file');
        assert.equal(data, 'firstsecond');
      }
      log('OK');
    }, 1000);
  }, 25);
});

```

Observed output:
```
11 Ready
18 Raw event info: change test.txt { watchedPath: 'workdir' }
21 Raw event info: change test.txt { watchedPath: 'workdir\\test.txt' }
22 Event on test.txt
23 Read first
24 Raw event info: change test.txt { watchedPath: 'workdir' }
25 Raw event info: change test.txt { watchedPath: 'workdir\\test.txt' }
45 Raw event info: change test.txt { watchedPath: 'workdir' }
46 Raw event info: change test.txt { watchedPath: 'workdir\\test.txt' }
1045 Never saw final file

AssertionError [ERR_ASSERTION]: 'first' == 'firstsecond'
```

**Expected behavior**
Any throttling needs to fire it's final event *after* the last throttled event occurred, not simply discard events that happen during the throttled time.

**Additional context**
I'm working on some automated tests for a build system, which often fail due to this, without artificial delays added, but this also explains errors I've had in the past in other projects using gulp-watch, where they processed an incomplete file that was only partially saved and never processed the final, correct file until I manually saved again.

This reproduces with any delay less than 49ms, so I'm assuming this is from the 50ms throttling I see in index.js around line 610.  I see the raw events are emitted reliably, should I just use those instead?  Maybe we can add an option to disable the throttling?

**Comment by beorn** [comment]
I'm running into this as well - very annoying.

How about replacing `throttle` with `throttle+debounce`, so the last event is also emitted?  That seems to be the common way to get around this.

See: <https://www.npmjs.com/package/throttle-debounce>

**Comment by Jimbly** [comment]
Well, after spending 3 hours debugging a build pipeline issue, only to find that the root cause was a bug I had debugged and tracked down years ago, I've implemented a workaround.  Totally hacky, but this seems to work, simply disables throttling of the 'change' events:
```javascript
function fixBuggyChokidar(watcher) {
  let orig = watcher._throttle.bind(watcher);
  watcher._throttle = function (actiontype, filename, timeout) {
    if (actiontype !== 'change') {
      return orig(actiontype, filename, timeout);
    }
    return true;
  };
}
```

Just call fixBuggyChokidar(watcher) and cross your fingers 🤞 .  Would love to see an actual fix land, but I can't quite decipher what's going on with the throttling myself or figure out how one would go about changing it to not drop the final event -_-.

Presumably performance will be worse with this, but worse performance is way preferable to not being notified of legitimate changes.  In my case, I managed to get a 99% reliably reproduction just saving a particular file in Photoshop (must have been just the right size that it didn't finish in time for the first read to pick it up, but fast enough to completely finish before the throttling released).
